src/wrapper/jnienv.rs

```diff
@@ -1932,6 +1967,202 @@ impl<'a> JNIEnv<'a> {
+        &self,
+        array: jarray,
+        is_copy: &mut jboolean,
+    ) -> Result<*mut c_void> {
+        non_null!(array, "get_primitive_array_critical array argument");
+        let res = jni_non_void_call!(self.internal, GetPrimitiveArrayCritical, array, is_copy);
+        Ok(res)
+    }
+
+    /// Release elements of the given array.
+    ///
+    /// See get_primitive_array_critical for a discussion on how these functions can be used.
+    pub fn release_primitive_array_critical(
+        &self,
+        array: jarray,
+        elems: &mut c_void,
+        mode: ReleaseMode,
+    ) -> Result<()> {
+        non_null!(array, "release_primitive_array_critical array argument");
+        jni_void_call!(
+            self.internal,
+            ReleasePrimitiveArrayCritical,
+            array,
+            elems,
+            mode as i32
+        );
+        Ok(())
+    }
+
+    /// Commit elements of the given primitive array.
+    ///
+    /// This function has no effect if elems is not a copy of the elements in array. Otherwise,
+    /// this function copies back the content of the array (and does not free the elems buffer).
+    pub fn commit_primitive_array_critical(
+        &self,
+        array: jbyteArray,
+        elems: &mut c_void,
+    ) -> Result<()> {
+        non_null!(array, "commit_primitive_array_critical array argument");
+        jni_void_call!(
+            self.internal,
+            ReleasePrimitiveArrayCritical,
+            array,
+            elems,
+            jni_sys::JNI_COMMIT
+        );
+        Ok(())
+    }
+
+    /// Return an AutoPrimitiveArray of the given Java primitive array.
+    ///
+    /// The result is valid until the corresponding AutoPrimitiveArray object goes out of scope,
+    /// when the release happens automatically according to the mode parameter.
+    ///
+    /// Given that Critical sections must be as short as possible, and that they come with a
+    /// number of important restrictions (see get_primitive_array_critical), use this
+    /// wrapper wisely, to avoid holding the array longer that strictly necessary.
+    /// In any case, you can:
+    ///  - Use the manual variants instead (i.e. get/release_primitive_array_critical).
+    ///  - Use std::mem::drop explicitly, to force / anticipate resource release.
+    ///  - Use a nested scope, to release the array at the nested scope's exit.
+    ///
+    /// Since the returned array may be a copy of the Java array, changes made to the
+    /// returned array will not necessarily be reflected in the original array until
+    /// release_primitive_array_critical() is called, which happens at AutoPrimitiveArray
+    /// destruction.
+    /// AutoPrimitiveArray has a commit() method, to force a copy of the array if needed (without
+    /// releasing it).
+    ///
+    /// See also [`get_primitive_array_critical`](struct.JNIEnv.html#method.get_primitive_array_critical)
+    pub fn get_primitive_array_critical_auto(
```

Comment:
I'm slightly worried about the naming inconsistency: we have `auto_local` and `get_primitive_array_critical_auto`.

Additionally, in Rust `get_` prefix is usually omitted, but I think that these arguments aren't very relevant in our case, because this library already has a lot functions with such prefix.

Personally, I would prefer to rename this function to `auto_primitive_array_critical` (or `get_auto_primitive_array_critical`) and use `AutoPrimitiveArray::new()` directly, so the `auto_primitive_array` function can be removed. Same for other `_auto` functions.

Reply:
OK.